### DISCUSSION TITLE
Adding more slash functionality

### DISCUSSION THREAD

**Discussion by lgarlati** [seerge/g-helper#2680] [Q&A] [answered]
Hello all,

I'm trying to add some functionality to the slash lighting for the 2024 models. However, I'm having some trouble understanding the structure and was wondering if there are any resources or files I can get pointed towards to start working on this. Specifically:
- How to set a slash pattern manually. I found this discussion (https://github.com/seerge/g-helper/discussions/2046#discussioncomment-9233416) but would appreciate a bit more explanation how to properly create and add a new pattern.
   - used "set custom", seems to work.
- How to locally build the project
    - Used .net and built with following command, am I missing any flags? `dotnet publish GHelper.sln --self-contained -p:PublishSingleFile=true`
- Does the slash support more than 3 brightness levels? 
- How to get current battery level % somewhere at or within the app/AnimeMatrix folder and subdirectories.

For those curious, I'm trying to add two main things:
- Display battery % on the slash
- "dim on battery" or simply have two distinct brightness levels. 
- "auto dim" where the brightness of the slash reflects the brightness of the screen. This way, when the screen is dimmed at night, the slash also dims accordingly.

**Comment by seerge** [comment, marked as answer]
@lgarlati hello,

> How to set a slash pattern manually. I found this discussion (https://github.com/seerge/g-helper/discussions/2046#discussioncomment-9233416) but would appreciate a bit more explanation how to properly create and add a new pattern.

I don't have a device by myself. But most probably each 4 bits (i.e. 1 letter) in a payload  ``"FF-FF-FF-FF-FF-FF-FF"`` is a brightness of a couple of LEDs 

>How to locally build the project

I build via VS GUI directly . But you can find build params for command line in github actions
https://github.com/seerge/g-helper/blob/main/.github/workflows/release.yml

> Does the slash support more than 3 brightness levels?

I does, but I try to keep things simple and unified with Anime Matrix controls (that support only 3).

> How to get current battery level % somewhere at or within the app/AnimeMatrix folder and subdirectories.

You get it from [windows WMI](https://stackoverflow.com/questions/8945986/find-out-battery-charge-capacity-in-percentage-using-c-sharp-or-net)

---

> "dim on battery" or simply have two distinct brightness levels.

What is the practical purpose of this ? Sash consumes battery power :) that's why both Anime Matrix and Slash Lighting control has an option to turn it off completely.

> "auto dim" where the brightness of the slash reflects the brightness of the screen. This way, when the screen is dimmed at night, the slash also dims accordingly.

You can control Slash Brightness with similar hotkeys to brightness already
> ``Fn + Shift + F7 / F8`` - Matrix / Slash Lighting brightness Down / Up

**Reply by lgarlati** [reply]
Yep, figured out setting the slash. Though it seems a bit wonkey, need to do more testing after work. 

> I build via VS GUI directly . But you can find build params for command line in github actions
https://github.com/seerge/g-helper/blob/main/.github/workflows/release.yml

Will look into this later, thank you. 

> It does, but I try to keep things simple and unified with Anime Matrix controls (that support only 3).

Fair enough, thanks.

> You get it from [windows WMI](https://stackoverflow.com/questions/8945986/find-out-battery-charge-capacity-in-percentage-using-c-sharp-or-net)

Thank you!

> What is the practical purpose of this ? Slash consumes battery power :) that's why both Anime Matrix and Slash Lighting control has an option to turn it off completely.

For sure. Off is less power draw than dim, no question there. I also agree that running patterns on it is purely cosmetic, and dimming would just make the cosmetic harder to see, so even less of a point. But, I'd still like to have it on, dimly, when on battery without having to change it manually every time. I also see why this may not be worth it simply for the clutter it would add to the otherwise clean UI.

> You can control Slash Brightness with similar hotkeys to brightness already

Yeah, but automation is the spice of life ;P

**Reply by lgarlati** [reply]
Managed to add the battery pattern. Had to use a timer to update the pattern every minute.

Going to work on the "auto dim" tomorrow
